File: client/lib/connection/impl/sessionkeypacket.dart
import 'dart:typed_data';

import 'package:client/connection/connection.dart';
import 'package:client/connection/packet.dart';
import 'package:client/crypto/aes.dart';

class SessionKeyPacket extends Packet {
  Key? _key;

  SessionKeyPacket.write(this._key);
  SessionKeyPacket.read() : _key = null;

  @override
  int getId() {
    return 1;
  }

  @override
  Uint8List write() {
    return _key!.serialize();
  }

  @override
  void read(Uint8List data) {
    _key = Key.deserialize(data);
  }

  @override
  void handle(Connection connection) {
    connection.setDecryptionCipher(AES(_key!));
    connection.setState(2);
    connection.eventBus.onSecureConnectionEstablished(connection);
  }
}

File: server/Server/src/me/spencernold/server/connection/packet/impl/SessionKeyPacket.java
package me.spencernold.server.connection.packet.impl;

import me.spencernold.server.connection.Client;
import me.spencernold.server.connection.Client.State;
import me.spencernold.server.connection.packet.Id;
import me.spencernold.server.connection.packet.Packet;
import me.spencernold.server.encryption.Key;
import me.spencernold.server.util.ByteBuffer;

@Id(id = 1)
public class SessionKeyPacket implements Packet {

	private Key sessionKey;
	
	public SessionKeyPacket(Key sessionKey) {
		this.sessionKey = sessionKey;
	}
	
	public SessionKeyPacket() {
	}
	
	public byte[] write() {
		return sessionKey.getBytes();
	}

	public void read(ByteBuffer buf) throws Exception {
		sessionKey = new Key(buf.toByteArray());
	}

	public void handle(Client client) {
		client.data.setDecryptionKey(sessionKey);
		Key key = Key.generate();
		client.data.setEncryptionKey(key);
		client.send(new SessionKeyPacket(key));
		client.state = State.SECURE;
	}
}
